blockchain/blockchain.go

```diff
@@ -0,0 +1,198 @@
+package blockchain
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/core/state"
+	"github.com/NethermindEth/juno/db"
+	"github.com/NethermindEth/juno/utils"
+)
+
+type ErrIncompatibleBlockAndStateUpdate struct {
+	reason string
+}
+
+func (e ErrIncompatibleBlockAndStateUpdate) Error() string {
+	return fmt.Sprintf("incompatible block and state update: %v", e.reason)
+}
+
+type ErrIncompatibleBlock struct {
+	reason string
+}
+
+func (e ErrIncompatibleBlock) Error() string {
+	return fmt.Sprintf("incompatible block: %v", e.reason)
+}
+
+// BlockDbKey creates appends block number and hash to create a db key.
+// Todo: make private after Blockchain.BlockByNumber & BlockchainBlockByHash are implemented.
+type BlockDbKey struct {
+	Number uint64
+	Hash   *felt.Felt
+}
+
+func (k *BlockDbKey) MarshalBinary() ([]byte, error) {
+	var numB [8]byte
+	binary.BigEndian.PutUint64(numB[:], k.Number)
+	return db.Blocks.Key(numB[:], k.Hash.Marshal()), nil
+}
+
+func (k *BlockDbKey) UnmarshalBinary(data []byte) error {
+	if len(data) != 41 {
+		return errors.New("key should be 41 bytes long")
+	}
+
+	if data[0] != byte(db.Blocks) {
+		return errors.New("wrong prefix")
+	}
+
+	k.Number = binary.BigEndian.Uint64(data[1:9])
+	k.Hash = new(felt.Felt).SetBytes(data[9:41])
+	return nil
+}
+
+// Blockchain is responsible for keeping track of all things related to the StarkNet blockchain
+type Blockchain struct {
+	l *sync.RWMutex
+
+	head     *core.Block
```

Comment:
For simplicity, this should be in the database. So that we can easily prove consistency of our database by using DB transactions. If not, our DB and memory could be out of sync.

Reply:
it already is: `HeadBlock`. I will just remove this reference.

Reply:
Done